lib_nn/api/MemCpyFn.hpp

```diff
@@ -0,0 +1,289 @@
+   * 0 (for dense convolutions) or 1 (for depthwise convolutions). This approach
+   * will also generalize if we ever end up with convolutions in which, for
+   * example, output channel 0 is a function of both input channels 0 and 1 (but
+   * no others), and output channel 1 is a function of both input channels 2 and
+   * 3 (but no others), etc, by using a channel stride value of 2.
+   *
+   * @param [inout] T   pointer to patch buffer
+   * @param [in] X      pointer to base address of the input image
+   * @param [in] h      row index of the _OUTPUT_ pixel being processed
+   * @param [in] w      column index of the _OUTPUT_ pixel being processed
+   * @param [in] c      channel index of the _INPUT_ image from which to start
+   * copying elements
+   */
+  virtual int8_t *memcopy_fn(int8_t *T, int8_t *X, int32_t h, int32_t w,
+                             int32_t c = 0) = 0;
+
+  /**
+   * @brief Get the number of bytes required for the scratch memory to hold the
+   * result of the mem copy. This accounts for the VPU memory read and write
+   * limitations.
+   *
+   * @return int
+   */
+  virtual int get_scratch_bytes() = 0;
+
+  /**
+   * @brief Get the number of bytes that VPU will overread from source tensor.
+   * This is to account for vector reads being used where a sub-vector read is
+   * required. This is used to prevent OOB memory accesses, i.e. if a memory
+   * location within this much beyond the end of an input tensor is invalid then
+   * it should be handled gracefully.
+   *
+   * @return int
+   */
+  virtual int get_overread_bytes() = 0;
+};
+
+/**
+ *
+ */
+class DerefInputFn : public MemCpyFn {
+ public:
+  class Params {
+   public:
+    int32_t bytes_per_h_line;
+    int32_t bytes_per_pixel;
+
+    /**
+     * @brief Construct a new Params object
+     *
+     * @param bytes_per_h_line Count of bytes in a horizontal line of the input
+     * tensor.
+     * @param bytes_per_pixel Count of bytes in a pixel of the input tensor.
+     */
+    Params(int32_t bytes_per_h_line, int32_t bytes_per_pixel);
+
+    /**
+     * @brief Construct a new Params object
+     *
+     * @param X Class describing the properties of the input tensor over which
+     * the convolution will be performed over.
+     * @param K Class describing the properties of the convolution to be
+     * performed.
+     */
+    Params(const ImageGeometry &X, const WindowGeometry &K);
+
+    /**
+     * @brief Construct a new Params object
+     *
+     * @param filter_geometry Class representing the properties of the input
+     * tensor and convolution properties over which the convolution will be
+     * performed.
+     */
+    Params(const Filter2dGeometry &filter_geometry);
+
+    Params(std::istream &stream);
+
+    void Serialize(std::ostream &stream) const;
+  };
+  /**
+   * @brief This describes the region over which this class will perform its
+   * operation(Memcopy).
+   */
+  Params *params;
```

Comment:
The `Params *params;` for this class and the following two classes, should they be exposed or be private members?

Reply:
good spot, I've made them private